File: Samples/State Pattern/Sample_0 - CPP/Player.cpp
#include "Player.h"

Player::~Player()
{
	delete m_currentState;
}

void Player::Update() {
	m_currentState->Execute();
}

void Player::SetState(State* new_state)
{
	if(m_currentState)
		m_currentState->Exit();

	delete m_currentState;

	m_currentState = new_state;
	m_currentState->Enter();
}

void Player::StateDisplay()
{
	cout << "[Player State]: " << m_currentState->GetName() << endl;
}

File: Samples/State Pattern/Sample_0 - CS/Player.cs
namespace Sample
{
    internal class Player
    {
        private State _currentState = null;

        public void Update()
        {
            _currentState.Execute();
        }

        public void SetState(State new_state)
        {
            if (_currentState != null)
                _currentState.Exit();

            _currentState = new_state;
            _currentState.Enter();
        }
    }
}
